File: December2021/C++/Populating_Next_Right_Pointers_in_Each_Node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if(!root) return root;
        queue<Node*> q;
        q.push(root);
        while(q.size())
        {
            int size=q.size();
            while(size>0)
            {
                Node* top=q.front();
                q.pop();
                if(top->left && top->right)
                {
                    q.push(top->left);
                    q.push(top->right);
                }
                if(--size==0) break;
                top->next=q.front();
            }
        }
        return root;
    }
};

File: December2021/PopulatingNextRightPointersInEachNode.java

class Solution {
    public Node connect(Node root) {

        if(root==null) {
            return null;
        }

        Queue<Node> qu = new LinkedList<>();
        qu.offer(root);



        while(!qu.isEmpty()) {
            int size = qu.size();
            Node prev = null;
            while(size-->0){
                Node head = qu.poll();
                // System.out.println(head.val +  " "+ size);
                head.next = prev;
                prev = head;
                if(head.right != null){
                    qu.offer(head.right);
                }
                if(head.left != null){
                    qu.offer(head.left);
                }

            }
        }
        //System.out.println("Hree");
        return root;
    }
}